pandas/io/formats/csvs.py

```diff
@@ -222,133 +303,63 @@ def save(self) -> None:
-            if encoded_labels and set(encoded_labels) != {""}:
-                encoded_labels.extend([""] * len(columns))
-                writer.writerow(encoded_labels)
-
     def _save(self) -> None:
-        self._save_header()
+        if self._need_to_save_header:
+            self._save_header()
+        self._save_body()
 
+    def _save_header(self) -> None:
+        if not self.has_mi_columns or self._has_aliases:
+            self.writer.writerow(self.encoded_labels)
+        else:
+            for row in self._generate_multiindex_header_rows():
+                self.writer.writerow(row)
+
+    def _generate_multiindex_header_rows(self) -> Generator[List[str], None, None]:
+        columns = self.obj.columns
+        for i in range(columns.nlevels):
+            # we need at least 1 index column to write our col names
+            col_line = []
+            if self.index:
+                # name is the first column
+                col_line.append(columns.names[i])
+
+                if isinstance(self.index_label, list) and len(self.index_label) > 1:
+                    col_line.extend([""] * (len(self.index_label) - 1))
+
+            col_line.extend(columns._get_level_values(i))
+            yield col_line
+
+        # Write out the index line if it's not empty.
+        # Otherwise, we will print out an extraneous
+        # blank line between the mi and the data rows.
+        if self.encoded_labels and set(self.encoded_labels) != {""}:
+            yield self.encoded_labels + [""] * len(columns)
+
+    def _save_body(self) -> None:
         nrows = len(self.data_index)
-
-        # write in chunksize bites
-        chunksize = self.chunksize
-        chunks = int(nrows / chunksize) + 1
-
+        chunks = int(nrows / self.chunksize) + 1
         for i in range(chunks):
-            start_i = i * chunksize
-            end_i = min((i + 1) * chunksize, nrows)
+            start_i = i * self.chunksize
+            end_i = min(start_i + self.chunksize, nrows)
             if start_i >= end_i:
                 break
-
             self._save_chunk(start_i, end_i)
 
     def _save_chunk(self, start_i: int, end_i: int) -> None:
-        data_index = self.data_index
+        ncols = self.obj.shape[-1]
+        data = [None] * ncols
 
         # create the data for a chunk
         slicer = slice(start_i, end_i)
 
         df = self.obj.iloc[slicer]
-        blocks = df._mgr.blocks
-
-        for i in range(len(blocks)):
-            b = blocks[i]
-            d = b.to_native_types(
-                na_rep=self.na_rep,
-                float_format=self.float_format,
-                decimal=self.decimal,
-                date_format=self.date_format,
-                quoting=self.quoting,
-            )
 
-            for col_loc, col in zip(b.mgr_locs, d):
-                # self.data is a preallocated list
-                self.data[col_loc] = col
+        for block in df._mgr.blocks:
```

Comment:
cc @jbrockmendel this could use block.apply in the future

Reply:
yah i think this and one other place in the pytables code.  i'll take another swing at these today